src/features/Dashboard/components/helpers.ts

```diff
@@ -97,3 +99,28 @@ export const constructTxTypeConfig = ({ type, protocol }: ITxTypeMeta): ITxTypeC
     }
   }
 });
+
+export const deriveDisplayAsset = (
+  txType: TxType,
+  to: TAddress,
+  networkId: NetworkId,
+  chainId: number,
+  valueTransfers: IFullTxHistoryValueTransfer[],
+  getAssetByContractAndNetworkId: (contractAddr: TAddress, networkId: NetworkId) => ExtendedAsset | undefined
+): Asset | undefined => {
+  switch (txType) {
+    default:
+      if (valueTransfers.length == 1) {
+        return valueTransfers[0].asset
+      }
+      return undefined;
+    case 'ERC_20_APPROVE':
+    case 'ERC_20_TRANSFER':
+    case ITxType.APPROVAL:
+      return getAssetByContractAndNetworkId(to, networkId) || generateGenericERC20(to, chainId.toString(), networkId)
+    case 'ERC_721_APPROVE':
+    case 'ERC_721_TRANSFER':
+    case 'ERC_721_MINT':
+      return generateGenericERC721(to, chainId.toString(), networkId)
```

Comment:
Test

Reply:
Moved this to `src/utils/generateAsset` and tested.